### PR TITLE
[SERV-543 Et Alia] CRUD Operations for Jobs

### PR DESCRIPTION
**Pull request by DRickard**
* Coded Add/get/update/delete methods for harvest jobs
* Utility methods for tweaking data for db format
* Related Tests
* Logging messages

### CODE REVIEW

File: src/test/java/edu/ucla/library/prl/harvester/utils/TestUtils.java [hunk 1]
```diff
@@ -63,4 +80,37 @@ public static Institution getRandomInstitution()
                 randWebsite);
     }
 
+    /**
+     * Gets a random {@link Job} for testing.
+     *
+     * @return A random Institution object
```
Comment (markmatney) [inline src/test/java/edu/ucla/library/prl/harvester/utils/TestUtils.java:86]:
```suggestion
     * @return A random Job object
```

File: src/test/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceIT.java [hunk 2]
```diff
@@ -217,4 +223,135 @@ public final void testUpdateInstitution(final Vertx aVertx, final VertxTestConte
...
+                            assertTrue(updated.getRepositoryBaseURL().equals(modified.getRepositoryBaseURL()));
+                        }).completeNow();
+                    }).onFailure(aContext::failNow);
+                }).onFailure(aContext::failNow);
+            } catch (MalformedURLException details) {
+                aContext.failNow(details);
+            }
+        }).onFailure(aContext::failNow);
+    }
+
```
Comment (markmatney) [inline src/test/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceIT.java:356]: Re: my comment about the update query failing if the institutionID was changed, there should be a test verifying that behavior.

**Review by markmatney — CHANGES_REQUESTED** [review]

File: src/main/java/edu/ucla/library/prl/harvester/Job.java [hunk 3]
```diff
@@ -26,17 +26,17 @@ public final class Job {
     /**
      * The JSON key for the institution ID.
      */
-    static final String INSTITUTION_ID = "institutionId";
+    static final String INSTITUTION_ID = "institutionid";
```
[the remark this answers is not in the document]
Comment (DRickard) [reply src/main/java/edu/ucla/library/prl/harvester/Job.java:29]: ETA: using camel-case names leads to errors like `io.vertx.serviceproxy.ServiceException: Missing value for field 'institutionId'`

[the remark this answers is not in the document]
Comment (ksclarke) [reply src/main/java/edu/ucla/library/prl/harvester/Job.java:29]: Yes, I was noticing that earlier and though about suggesting we change to snake case.

File: src/main/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceImpl.java [hunk 4]
```diff
@@ -187,6 +223,34 @@ private String getOptionalPhoneAsString(final Optional<PhoneNumber> aPhoneParam)
...
+
+    /**
+     * Converts Optional zoned timestamp value to offset time value for use in prepared queries.
+     *
+     * @param aTimeParam An Optional zoned timestamp used as a query param
+     * @return The OffsetDateTime representation of the Optional value, or a null if Optional is empty
+     */
+    private OffsetDateTime getOptionalTimeAsOffset(final Optional<ZonedDateTime> aTimeParam) {
+        if (aTimeParam.isPresent()) {
+            return OffsetDateTime.from(aTimeParam.get());
```
[the remark this answers is not in the document]
Comment (DRickard) [reply src/main/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceImpl.java:248]: Don't know if it's Postgres or the vert.x Postgres library, but yes one/both of them demanded the OffsetDateTime encoding.  Created SERV-599 for the type change.

File: src/main/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceImpl.java [hunk 5]
```diff
@@ -216,32 +280,74 @@ public Future<Void> removeInstitution(final Integer anInstitutionId) {
...
+        return myDbConnectionPool.withConnection(connection -> {
+            return connection.preparedQuery(GET_JOB).execute(Tuple.of(aJobId));
+        }).recover(error -> {
+            return Future.failedFuture(new ServiceException(INTERNAL_ERROR, error.getMessage()));
+        }).compose(select -> {
+            if (hasSingleRow(select)) {
+                return Future.succeededFuture(new Job(select.iterator().next().toJson()));
+            }
+            return Future.failedFuture(
+                    new ServiceException(NOT_FOUND_ERROR, LOGGER.getMessage(MessageCodes.PRL_007, aJobId)));
```
[the remark this answers is not in the document]
Comment (DRickard) [reply src/main/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceImpl.java:292]: D'oh! Cut-and-paste programming gets me every time...

File: src/test/java/edu/ucla/library/prl/harvester/utils/TestUtils.java (hunk 1 above)
Comment (DRickard) [reply src/test/java/edu/ucla/library/prl/harvester/utils/TestUtils.java:86]: D'oh...

File: src/main/java/edu/ucla/library/prl/harvester/Job.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (markmatney) [reply src/main/java/edu/ucla/library/prl/harvester/Job.java:29]: SERV-600 created to reflect decision made in meeting

File: src/main/java/edu/ucla/library/prl/harvester/Job.java [hunk 6]
```diff
@@ -26,17 +26,17 @@ public final class Job {
...
      * The JSON key for the repository base URL.
      */
-    static final String REPOSITORY_BASE_URL = "repositoryBaseUrl";
+    static final String REPOSITORY_BASE_URL = "repositorybaseurl";
 
     /**
      * JSON key for the metadata prefix.
      */
-    static final String METADATA_PREFIX = "metadataPrefix";
+    static final String METADATA_PREFIX = "metadataPpefix";
```
Comment (markmatney) [inline src/main/java/edu/ucla/library/prl/harvester/Job.java:39]:
```suggestion
    static final String METADATA_PREFIX = "metadataprefix";
```

File: src/test/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceIT.java [hunk 7]
```diff
@@ -354,4 +355,33 @@ public final void testUpdateJob(final Vertx aVertx, final VertxTestContext aCont
...
+    @Test
+    public final void testUpdateJobBadID(final Vertx aVertx, final VertxTestContext aContext)
+            throws AddressException, MalformedURLException, NumberParseException {
+        final int jobID = 1;
+        final int badInstID = -1;
+        myScheduleStoreProxy.getJob(jobID).onSuccess(original -> {
+            try {
+                final Job modified = new Job(badInstID, new URL("http://new.url.com"), original.getSets().get(),
+                        original.getScheduleCronExpression(), ZonedDateTime.now());
+                myScheduleStoreProxy.updateJob(jobID, modified).onSuccess(result -> {
```
Comment (markmatney) [inline src/test/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceIT.java:375]: Can you make it so updateJob fails if no rows were updated?

File: src/main/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceImpl.java [hunk 8]
```diff
@@ -73,6 +80,35 @@ public class HarvestScheduleStoreServiceImpl implements HarvestScheduleStoreServ
...
+     * The delete query for a job.
+     */
+    private static final String DEL_JOB = "DELETE FROM public.harvestjobs WHERE id = $1";
+
+    /**
+     * The update query for a job.
+     */
+    private static final String UPDATE_JOB =
+            "UPDATE public.harvestjobs SET repositoryBaseURL=$1, sets=$2, lastSuccessfulRun=$3," +
+                    " scheduleCronExpression=$4 WHERE id = $5";
```
[the remark this answers is not in the document]
Comment (DRickard) [reply src/main/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceImpl.java]: Added logic to test if update happened, and return failed future if no row (and by implication, more than one row) was touched.

File: src/test/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceIT.java (hunk 2 above)
Comment (DRickard) [reply src/test/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceIT.java:358]: Added test for failed update event,

File: src/test/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceIT.java (hunk 7 above)
Comment (DRickard) [reply src/test/java/edu/ucla/library/prl/harvester/services/HarvestScheduleStoreServiceIT.java]: Added failed future if no update happened

**Review by ksclarke — DISMISSED** [review]

File: src/main/java/edu/ucla/library/prl/harvester/Error.java [hunk 9]
```diff
@@ -25,11 +25,16 @@ public enum Error {
...
+     * The failure code if the database service runs a "get/update" query that finds no matching rows
      */
     NOT_FOUND,
 
+    /**
+     * The failure code if the database service runs a "get/update" query that finds 2+ matching rows when there should
+     * only be one
+     */
+    TOO_MANY_MATCHES,
+
```
Comment (markmatney) [inline src/main/java/edu/ucla/library/prl/harvester/Error.java:37]: Is this needed anymore? It isn't being referenced.

**Review by markmatney — COMMENTED** [review]
All else looks good, just one question.

File: src/main/java/edu/ucla/library/prl/harvester/Error.java (hunk 9 above)
Comment (DRickard) [reply src/main/java/edu/ucla/library/prl/harvester/Error.java]: Removed

**Review by markmatney — APPROVED** [review]

**Review by ksclarke — APPROVED** [review]

**Pull request merged**